avocado/core/spawners/common.py

```diff
@@ -20,6 +27,44 @@ class BaseSpawner:
 
     METHODS = []
 
+    def bytes_from_file(self, filename):
+        """Read bytes from a files in binary mode.
+
+        This is a helpful method to read *local* files bytes efficiently.
+
+        If the spawner that you are implementing needs access to local file,
+        feel free to use this method.
+        """
+        # This could be optimized in the future.
+        with open(filename, 'rb', 0) as fp:
+            with mmap(fp.fileno(), 0, access=ACCESS_READ) as stream:
+                yield stream.read()
+
+    def stream_output(self, job_id, task_id):
+        """Returns output files streams in binary mode from a task.
+
+        This method will find for output files generated by a task and will
+        return a generator with tuples, each one containing a filename and
+        bytes.
+
+        You need to provide in your spawner a `stream_output()` method if this
+        one is not suitable for your spawner. i.e: if the spawner is trying to
+        access a remote output file.
+        """
+        src = '~/avocado/job-results/{}/test-results/{}/data'.format(job_id,
```

Comment:
Is this intended as an example location (the default) or are you suggesting that we make it the only allowed location for a test runner to write to?

My point is that this location, if on the system that originated the *job*, should possibly also respect `run.results_dir`, right?

Also, from the runner perspective (let's imagine the job is running on a host, and the test runner on another), do you think it's necessary or a good idea to replicate the `~/avocado/job-results/...` structure? Or maybe use something that is volatile and temporary such as something under `/var/tmp` or `/tmp`?

Reply:
So, I went to double check and we're currently setting `AVOCADO_TEST_OUTPUT_DIR` to `/tmp/.avocado-task-xxxxxx`.  So this is my main question... how will a spawner, such as a process spawner, find that location so that it can be streamed back to the job?

Reply:
@clebergnu this was just a test, but for sure we need to respect the `run.results_dir`. Because today, `ProcessSpawner` is local and `PodmanSpawner` can also be local (in this case, mount directly the fs to the container).

The `~/avocado/job-results/` structure, IMO should be created by the `Job()`, so just once and at the job level and on the job machine. If the test is spawned on a remote machine, test should only save the test-results folder there. And the `RemoteSpawner` should be responsible for copying this only during the "download" process/command.

IMO, the `Job()` should save inside the `~/avocado/job-results/latest/jobdata/results.json` or `~/avocado/job-results/latest/results.json` the spawner that has executed this job. On top of that, each test should save (on the 'started' or 'finished' status) the OUTPUT_DIR that was used.

Makes sense?

Reply:
> @clebergnu this was just a test, but for sure we need to respect the `run.results_dir`. Since that today, `ProcessSpawner` is local and `PodmanSpawner` can also be local (in this case, mount directly the fs to the container) I assumed that the simplest implementation would be local.
> 

OK, fair enough.

One extra point is that we need to understand if this mapping for a `PodmanSpawner` needs or makes sense to be symmetrical, I mean, also be available from the container PoV as "~/avocado/job-results/job-xxxxxxxxxxxx/test-results/yyyyyyyyyy/". My initial reaction is that it should be a simpler location on the container side, but I don't have hard facts to back it, just a gut feeling at this point ;)

> The `~/avocado/job-results/` structure, IMO should be created by the `Job()`, so just **once**, at the job level and on the **job machine**. If the test is spawned on a remote machine, test should only save the test-results folder there. And the `RemoteSpawner` should be responsible for copying this only during the "download" process/command.
> 

Ack, makes complete sense to me.  I'd just like to clarify that the structure that the job would be (actually is) responsible to create includes extra components, such as the `job-YYYY-MM-DDTHH.MM-$shortid/test-results`, but that remains the same.

> IMO, the `Job()` should save inside the `~/avocado/job-results/latest/jobdata/results.json` or `~/avocado/job-results/latest/results.json` the spawner that has executed this job. On top of that, each test should save (on the 'started' or 'finished' status) the OUTPUT_DIR that was used.
> 

That's an interesting point to be discussed.  I still don't have an opinion on how much job related information should be kept in the job, or on a task.  Right now we have the "recipes", the JSON serialization, that in theory could hold information themselves. Examples:

* When a user runs a job and provides a global parameter (via any means, such as a varianter) setting `network` to `bridge`, assuming the parameter will be passed to tests (Tasks), should that parameter value be preserved in the Task serialization so that the Task (and the job) can be replayed?

* When a user replays the same job, should the Task serialization keep information about the spawner that was used in its original execution? This may be important because often a user will want another user to replay a single task (say a failing one), and we may be deciding on the granularity level that people will share information about reproducing tests executions.

> Makes sense?

It does, and it generates a few questions that I believe we'll find the best answer when the time is right.